pytube/contrib/search.py

```diff
@@ -0,0 +1,198 @@
+"""Module for interacting with YouTube search."""
+# Native python imports
+import logging
+
+# Local imports
+from pytube import YouTube
+from pytube.innertube import InnerTube
+
+
+logger = logging.getLogger(__name__)
+
+
+class Search:
+    def __init__(self, query):
+        self.query = query
+        self._innertube_client = InnerTube()
+
+        self._initial_results = None
+        self._results = None
+        self._completion_suggestions = None
+        self._current_continuation = None
+
+    @property
+    def completion_suggestions(self):
+        """Return query autocompletion suggestions for the query.
+
+        :rtype: list
+        :returns:
+            A list of autocomplete suggestions provided by YouTube for the query.
+        """
+        if self._completion_suggestions:
+            return self._completion_suggestions
+        if self.results:
+            self._completion_suggestions = self._initial_results['refinements']
+        return self._completion_suggestions
+
+    @property
+    def results(self):
+        """Return search results.
+
+        On first call, will generate and return the first set of results.
+        Additional results can be generated using ``.get_next_results()``.
+
+        :rtype: list
+        :returns:
+            A list of YouTube objects.
+        """
+        if self._results:
+            return self._results
+
+        videos, continuation = self.fetch_and_parse()
+        self._results = videos
+        self._current_continuation = continuation
+        return self._results
+
+    def get_next_results(self):
+        """Use the stored continuation string to fetch the next set of results.
+
+        This method does not return the results, but instead updates the results property.
+        """
+        if self._current_continuation:
+            videos, continuation = self.fetch_and_parse(self._current_continuation)
+            self._results.extend(videos)
+            self._current_continuation = continuation
+        else:
+            raise IndexError
```

Comment:
maybe can return StopIteration instead similar to iterator?

Reply:
`StopIteration` only really makes sense if this were implemented as an iterator, see above for my argument against. `IndexError` probably isn't the best either though, might be worth adding a new pytube exception here

Reply:
Generator should also return StopIteratioon I believe? https://stackoverflow.com/questions/2322342/should-a-python-generator-raise-an-exception-when-there-are-no-more-elements-to

Reply:
Generators aren't supposed to raise StopIteration themselves (https://www.python.org/dev/peps/pep-0479/). Instead you just return.

Reply:
Ah oops - i'm a bit outdated there :-)